app/questionnaire/router.py

```diff
@@ -205,19 +216,28 @@ def get_previous_location_url(
     def _get_return_to_location_url(
         self,
         location: Location,
+        is_section_complete: bool,
         return_to: str,
         return_to_answer_id: Optional[str] = None,
-    ) -> Optional[str]:
-        if return_to == "section-summary":
+        return_to_block_id: Optional[str] = None,
+    ) -> str:
+        if return_to == "section-summary" and is_section_complete:
             return self._get_section_url(
                 location, return_to_answer_id=return_to_answer_id
             )
 
-        if return_to == "final-summary" and self.is_questionnaire_complete:
+        if (
+            return_to == "final-summary"
+            and is_section_complete
+            and self.is_questionnaire_complete
+        ):
             return url_for(
                 "questionnaire.submit_questionnaire", _anchor=return_to_answer_id
             )
 
+        if return_to == "calculated-summary" and return_to_block_id:
+            return url_for("questionnaire.block", block_id=return_to_block_id)
```

Comment:
These `if` statements will need to be nested. i.e if return to is `"section-summary"` but section isn't complete, we shouldn't evaluate the next if and so on. We should know at that point to return None.

Adding on to my previous comment, it could be something like:

```py
        if not return_to:
            return None

        if return_to == "calculated-summary":
            if return_to_block_id:
                return url_for("questionnaire.block", block_id=return_to_block_id)

            return None

        if is_section_complete is None:
            is_section_complete = self._progress_store.is_section_complete(
                location.section_id, location.list_item_id
            )

        if not is_section_complete:
            return None

        if return_to == "section-summary":
            return self._get_section_url(
                location, return_to_answer_id=return_to_answer_id
            )
        if return_to == "final-summary" and self.is_questionnaire_complete:
            return url_for(
                "questionnaire.submit_questionnaire", _anchor=return_to_answer_id
            )
```
I agree it is a lot of `ifs` but unavoidable without a redesign.